x509-cert/src/ext/pkix/sct.rs

```diff
@@ -0,0 +1,691 @@
+#![cfg(feature = "sct")]
+//! Signed Certificate Timestamp list extension as defined in the
+//! [Certificate Transparency RFC 6962].
+//!
+//! [Certificate Transparency RFC 6962]: https://datatracker.ietf.org/doc/html/rfc6962
+
+use alloc::{format, vec::Vec};
+use const_oid::{AssociatedOid, ObjectIdentifier};
+use der::asn1::OctetString;
+use tls_codec::{
+    DeserializeBytes, SerializeBytes, TlsByteVecU16, TlsDeserializeBytes, TlsSerializeBytes,
+    TlsSize,
+};
+
+/// A signed certificate timestamp list (SCT list) as defined in [RFC 6962 Section 3.3].
+///
+/// ```text
+/// SignedCertificateTimestampList ::= OCTET STRING
+/// ```
+///
+/// [RFC 6962 Section 3.3]: https://datatracker.ietf.org/doc/html/rfc6962#section-3.3
+#[derive(Debug, PartialEq)]
+pub struct SignedCertificateTimestampList(OctetString);
+
+//TODO: Remove this and use const-oid's rfc6962::CT_PRECERT_SCTS once a const-oid version
+// containing it is published
+const CT_PRECERT_SCTS: ObjectIdentifier = ObjectIdentifier::new_unwrap("1.3.6.1.4.1.11129.2.4.2");
+
+impl AssociatedOid for SignedCertificateTimestampList {
+    const OID: ObjectIdentifier = CT_PRECERT_SCTS;
+}
+
+impl_newtype!(SignedCertificateTimestampList, OctetString);
+impl_extension!(SignedCertificateTimestampList, critical = false);
+
+/// Errors that are thrown by this module.
+#[derive(PartialEq, Debug)]
+pub enum Error {
+    /// [Errors][der::Error] from the `der` crate.
+    Der(der::Error),
+    /// [Errors][tls_codec::Error] from the `tls_codec` crate.
+    Tls(tls_codec::Error),
+}
+
+impl From<der::Error> for Error {
+    fn from(value: der::Error) -> Self {
+        Error::Der(value)
+    }
+}
+
+impl From<tls_codec::Error> for Error {
+    fn from(value: tls_codec::Error) -> Self {
+        Error::Tls(value)
+    }
+}
+
+impl SignedCertificateTimestampList {
+    /// Creates a new [SignedCertificateTimestamp] from a slice of [SerializedSct]s.
+    pub fn new(serialized_scts: &[SerializedSct]) -> Result<Self, Error> {
+        let mut result: Vec<u8> = Vec::new();
+        for timestamp in serialized_scts {
+            let buffer = timestamp.tls_serialize()?;
+            result.extend(buffer);
+        }
+        let tls_vec = TlsByteVecU16::new(result);
+        let buffer = tls_vec.tls_serialize()?;
+        Ok(SignedCertificateTimestampList(OctetString::new(buffer)?))
+    }
+
+    /// Parses the encoded [SerializedSct]s and returns a [Vec][alloc::vec::Vec] containing them.
+    ///
+    /// Returns an [error][Error] if a [SerializedSct] can't be
+    /// deserialized or if there are trailing bytes after all [SerializedSct]s
+    /// are deserialized.
+    pub fn parse_timestamps(&self) -> Result<Vec<SerializedSct>, Error> {
+        let (tls_vec, rest) = TlsByteVecU16::tls_deserialize(self.0.as_bytes())?;
```

Comment:
We renamed this version to `tls_deserialize_bytes` to disambiguate the two versions.